### PR TITLE
QRCode scanner

### PR DESCRIPTION
**Pull request by GabrielFleischer**
Add a qrcode scanner to the connect ui

### CODE REVIEW

**Review by haoqianzhang — APPROVED** [review]

File: app/src/main/java/com/github/dedis/student20_pop/qrcode/CameraPreview.java [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        }
+
+        @Override
+        public void surfaceChanged(@NonNull SurfaceHolder holder, int format, int width, int height) {
+        }
+    }
+
+    @Override
+    protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
+        int width = 240;
```
Comment (celinecamacho) [inline app/src/main/java/com/github/dedis/student20_pop/qrcode/CameraPreview.java:112]: Could you store this values in res/values/dimens.xml?

File: app/src/main/java/com/github/dedis/student20_pop/ui/ConnectFragment.java [hunk 2]
```diff
@@ -1,26 +1,143 @@
...
+
+        BarcodeDetector qrDetector = new BarcodeDetector.Builder(getContext())
+                .setBarcodeFormats(Barcode.QR_CODE)
+                .build();
+
+        qrDetector.setProcessor(new QRFocusingProcessor(qrDetector, this));
+
+        return new CameraSource.Builder(requireContext(), qrDetector)
+                .setFacing(CameraSource.CAMERA_FACING_BACK)
+                .setRequestedPreviewSize(1600, 1024)
```
Comment (celinecamacho) [inline app/src/main/java/com/github/dedis/student20_pop/ui/ConnectFragment.java:73]: Same comment for these values: 1600, 1024 and 15.0f

File: app/src/main/res/layout/fragment_connect.xml [hunk 3]
```diff
@@ -1,21 +1,78 @@
...
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        app:layout_constrainedHeight="true"
+        app:layout_constraintBottom_toBottomOf="parent"
+        app:layout_constraintTop_toBottomOf="@id/guideline_horizontal">
+
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_height="match_parent"
+            android:layout_margin="50dp">
```
Comment (celinecamacho) [inline app/src/main/res/layout/fragment_connect.xml:37]: Same comment for the values in line 37, 42 and 44

File: app/src/main/res/layout/fragment_connect.xml [hunk 4]
```diff
@@ -1,21 +1,78 @@
...
+                android:text="@string/allow_camera_text"
+                android:textSize="@dimen/size_body"
+                android:textStyle="bold" />
+
+            <Button
+                android:id="@+id/allow_camera_button"
+                style="@style/Widget.AppCompat.Button"
+                android:layout_width="wrap_content"
+                android:layout_height="wrap_content"
+                android:background="#03A9F4"
```
Comment (celinecamacho) [inline app/src/main/res/layout/fragment_connect.xml:70]: You can store the colors in the file res/values/colors.xml and then access them using @colors/id

File: app/src/main/res/values/strings.xml [hunk 5]
```diff
@@ -22,5 +26,6 @@
 
     <!-- Temp -->
     <string name="text_temp_connect">Connect UI</string>
```
Comment (celinecamacho) [inline app/src/main/res/values/strings.xml:28]: You can delete the two strings "text_temp_connect" and "text_temp_launch" if you're not using them anymore

**Review by celinecamacho — COMMENTED** [review]
Great job, just a few comments on values that should be stored in res/values

**Review by celinecamacho — APPROVED** [review]

File: app/src/main/java/com/github/dedis/student20_pop/qrcode/CameraPreview.java [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+
+import androidx.annotation.NonNull;
+import androidx.annotation.RequiresPermission;
+
+import com.google.android.gms.common.images.Size;
+import com.google.android.gms.vision.CameraSource;
+
+import java.io.IOException;
+
+public class CameraPreview extends ViewGroup {
```
Comment (lhmerino) [inline app/src/main/java/com/github/dedis/student20_pop/qrcode/CameraPreview.java:34]: Please add a comment to discuss what is the purpose of this class

File: app/src/main/java/com/github/dedis/student20_pop/qrcode/QRFocusingProcessor.java [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+import android.util.SparseArray;
+
+import com.google.android.gms.vision.Detector;
+import com.google.android.gms.vision.FocusingProcessor;
+import com.google.android.gms.vision.Tracker;
+import com.google.android.gms.vision.barcode.Barcode;
+import com.google.android.gms.vision.barcode.BarcodeDetector;
+
+public class QRFocusingProcessor extends FocusingProcessor<Barcode> {
```
Comment (lhmerino) [inline app/src/main/java/com/github/dedis/student20_pop/qrcode/QRFocusingProcessor.java:11]: Comment on this class too

File: app/src/androidTest/java/com/github/dedis/student20_pop/ui/ConnectFragmentTest.java [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+import org.junit.Test;
+
+import static androidx.test.espresso.Espresso.onView;
+import static androidx.test.espresso.action.ViewActions.click;
+import static androidx.test.espresso.assertion.ViewAssertions.matches;
+import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
+import static androidx.test.espresso.matcher.ViewMatchers.withId;
+import static androidx.test.espresso.matcher.ViewMatchers.withText;
+
+public class ConnectFragmentTest {
```
Comment (lhmerino) [inline app/src/androidTest/java/com/github/dedis/student20_pop/ui/ConnectFragmentTest.java:20]: Please add comment here as well

**Review by lhmerino — CHANGES_REQUESTED** [review]

**Comment by lhmerino** [comment]
Looks great, thank you! Just would like to see more comments to understand the purpose of each class at a quick glance.

**Review by lhmerino — APPROVED** [review]

**Pull request merged**